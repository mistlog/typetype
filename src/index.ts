export * from "./parser";
export * from "./generator";
export * from "./core";