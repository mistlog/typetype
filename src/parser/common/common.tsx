import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}