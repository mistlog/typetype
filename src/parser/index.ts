export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";