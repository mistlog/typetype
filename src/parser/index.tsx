export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";