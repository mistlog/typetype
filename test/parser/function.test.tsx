import { ReactPeg } from "react-peg";
import { TypeExpression } from "../../src";
import { saveAST } from "../common";

describe("parser: function", () => {
    test("FunctionType", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`type () => void`);
        saveAST(ast, "FunctionType.json");
        expect(ast).toMatchSnapshot();
    })

    test("FunctionType: rest", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`type (...args: any[]) => void`);
        saveAST(ast, "FunctionType-Rest.json");
        expect(ast).toMatchSnapshot();
    })

    test("FunctionType: params", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`type (a:number, b: string) => number`);
        saveAST(ast, "FunctionType-Params.json");
        expect(ast).toMatchSnapshot();
    })

    test("FunctionType: nested", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`type () => type (a: number, b: string) => void`);
        saveAST(ast, "FunctionType-Nested.json");
        expect(ast).toMatchSnapshot();
    })

    test("FunctionType: type param", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`type <K extends string, T>(key: K, value: T) => void`);
        saveAST(ast, "FunctionType-TypeParam.json");
        expect(ast).toMatchSnapshot();
    })
})