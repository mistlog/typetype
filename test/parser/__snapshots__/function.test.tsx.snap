// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: function FunctionType 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [],
  "returnType": Object {
    "kind": "VoidType",
    "value": "void",
  },
}
`;

exports[`parser: function FunctionType: nested 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [],
  "returnType": Object {
    "kind": "FunctionType",
    "params": Array [
      Object {
        "kind": "FunctionTypeParam",
        "name": Object {
          "kind": "Identifier",
          "name": "a",
        },
        "type": Object {
          "kind": "NumberType",
          "value": "number",
        },
      },
      Object {
        "kind": "FunctionTypeParam",
        "name": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "type": Object {
          "kind": "StringType",
          "value": "string",
        },
      },
    ],
    "returnType": Object {
      "kind": "VoidType",
      "value": "void",
    },
  },
}
`;

exports[`parser: function FunctionType: optional param 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "type": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "b",
      },
      "optional": true,
      "type": Object {
        "kind": "StringType",
        "value": "string",
      },
    },
  ],
  "returnType": Object {
    "kind": "NumberType",
    "value": "number",
  },
}
`;

exports[`parser: function FunctionType: params 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "type": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "b",
      },
      "type": Object {
        "kind": "StringType",
        "value": "string",
      },
    },
  ],
  "returnType": Object {
    "kind": "NumberType",
    "value": "number",
  },
}
`;

exports[`parser: function FunctionType: rest 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "args",
      },
      "rest": true,
      "type": Object {
        "dimension": 1,
        "elementType": Object {
          "kind": "AnyType",
          "value": "any",
        },
        "kind": "ArrayType",
      },
    },
  ],
  "returnType": Object {
    "kind": "VoidType",
    "value": "void",
  },
}
`;

exports[`parser: function FunctionType: type param 1`] = `
Object {
  "kind": "FunctionType",
  "params": Array [
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "key",
      },
      "type": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "K",
        },
      },
    },
    Object {
      "kind": "FunctionTypeParam",
      "name": Object {
        "kind": "Identifier",
        "name": "value",
      },
      "type": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "T",
        },
      },
    },
  ],
  "returnType": Object {
    "kind": "VoidType",
    "value": "void",
  },
  "typeParams": Array [
    Object {
      "constraint": Object {
        "kind": "StringType",
        "value": "string",
      },
      "kind": "TypeReference",
      "typeName": Object {
        "kind": "Identifier",
        "name": "K",
      },
    },
    Object {
      "kind": "TypeReference",
      "typeName": Object {
        "kind": "Identifier",
        "name": "T",
      },
    },
  ],
}
`;
